gt/main.c

```diff
@@ -292,8 +294,18 @@ gt_fill_up_ether_cache_locked(struct ether_cache *eth_cache,
 			ip_dst, sizeof(eth_cache->ip_addr.ip.v6));
 	}
 
-	eth_cache->eth_hdr.ether_type = rte_cpu_to_be_16(inner_ip_ver);
-	ether_addr_copy(&iface->eth_addr, &eth_cache->eth_hdr.s_addr);
+	if (iface->vlan_insert) {
+		eth_cache->l2_hdr.eth_hdr.ether_type =
+			rte_cpu_to_be_16(ETHER_TYPE_VLAN);
+		eth_cache->l2_hdr.vlan_hdr.vlan_tci = iface->vlan_tag_be;
+		eth_cache->l2_hdr.vlan_hdr.eth_proto =
+			rte_cpu_to_be_16(inner_ip_ver);
```

Comment:
If we move `fill_vlan_hdr()` from `gt/main.c` to `include/gatekeeper_l2.h`, we can replace the whole true-case body of the `if` statement here with the following call `fill_vlan_hdr(&eth_cache->l2_hdr.eth_hdr, iface->vlan_tag_be, inner_ip_ver);`.